### PR TITLE
CmdStanMCMC/MLE/VB from read_cmdstan_csv ouput

### LINKED ISSUE #411: Create CmdStanMCMC R6 object from CSV
**Is your feature request related to a problem? Please describe.**
Currently no sane way to create `CmdStanMCMC` object as returned by `$sample` command by reading from CSV. 

**Describe the solution you'd like**
Something like
```r
fit <- to.cmdstan.mcmc(read_cmdstan_csv(c(file1, file2, file3, file4)))
```
so that later `fit` can be used for various actions such as
```r
draws <- fit$draws()
```

**Describe alternatives you've considered**
None so far.

**Additional context**
One motivation is to help user share/debug outputs, as indicated [here](https://discourse.mc-stan.org/t/draws-method-in-cmdstanr-is-still-slow/19490/5?u=yizhang). Another one is to port past cmdstan output to be postprocessed.

### PR DESCRIPTION
**Pull request by jgabry**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

Closes #411 

Allows creating CmdStanMCMC, CmdStanVB, and CmdStanMLE objects from the list returned by `read_cmdstan_csv()`. 
In reality these objects are not exactly the same because they don't have an associated CmdStanRun object. This means that methods like `cmdstan_diagnose()` and `save_output_files()` are unavailable, but the most important methods (e.g., `draws()`, `summary()`) work fine. 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
**Columbia University**

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: R/fit.R
Comment (jgabry) [inline R/fit.R:20]: Need to remove this so it doesn't error if output files aren't available. Otherwise we can't use this with the output from read_cmdstan_csv(). 

However, this doesn't risk `print()` failing to catch an important error because the same error will be caught by the `metadata()` method that is called internally by `print()`. `metadata()` is smarter about throwing the error because it only throws the error if the files are missing AND `metadata()` hasn't already been read in.

File: R/fit.R
Comment (jgabry) [inline R/fit.R:330]: We need to check the method via metadata()$method and not runset$method() or it won't work with the objects created after read_cmdstan_csv

File: R/fit.R
Comment (jgabry) [inline R/fit.R:602]: Move this error after the `is.null` check so that it's only thrown if metadata hasn't already been read in.

File: R/fit.R
Comment (jgabry) [inline R/fit.R:705]: This and every other change to `csv_contents` isn't important, it just makes naming consistent throughout our code

File: R/fit.R
Comment (jgabry) [inline R/fit.R:733]: This error will also be thrown by `private$read_csv_` so it is unnecessary here. We also need to remove it or it will result in errors after creating the objects from read_cmdstan_csv.

File: R/fit.R
Comment (jgabry) [inline R/fit.R:770]: this all and cases below are just name changes

File: R/read_csv.R
Comment (jgabry) [inline R/read_csv.R:388]: What do you think of the names `as_cmdstan_mcmc()`, `as_cmdstan_mle()`, `as_cmdstan_vb()`?

File: R/read_csv.R
Comment (jgabry) [inline R/read_csv.R:418]: Don't actually create CmdStanMCMC objects but CmdStanMCMC2 objects that inherit from CmdStanMCMC. Same approach with MLE and VB below. This avoids having to create the runset object because we can override the `initialize` method.  

Suggestions for better class name than `CmdStanMCMC2`?

File: R/read_csv.R
Comment (jgabry) [inline R/read_csv.R:479]: This makes sure that the methods that won't work have informative error messages.

**Review by jgabry — COMMENTED** [review]
@rok-cesnovar Just added a bunch of review comments that hopefully help explain the changes.

File: DESCRIPTION
Comment (jgabry) [inline DESCRIPTION:3]: bumping to dev version

File: R/read_csv.R
Comment (jgabry) [inline R/read_csv.R:429]: I put these in the read_csv.R file because of the use case but we could also but these in the fit.R file with the definition of CmdStanMCMC, CmdStanMLE, etc. Preference?

**Comment by jgabry** [comment]
@yizhang-yiz I think this is now ready to play around with if you want to try it out.

**Comment by yizhang-yiz** [comment]
```r
> fit <- cmdstanr::as_cmdstan_mcmc(dir(pattern=".csv", full.name=TRUE))
```
This works
```r
bayesplot::mcmc_dens_overlay(fit$draws(), pars = c("CL", "Q", "V1", "V2", "ka", "sigma"), facet_args = list(nrow = 2))
```
but this doesn't
```r
> print(fit$time(), digits = 3)
Error: This method is not available for objects created using as_cmdstan_mcmc().
```
I wonder what the rationale is.

**Comment by jgabry** [comment]
> I wonder what the rationale is.

No good one other than it would take more effort to get that to work. `read_cmdstan_csv()` doesn't currently give you the timing information. When model fitting happens we create a CmdStanRun object

https://github.com/stan-dev/cmdstanr/blob/46ae80f12881356e9152edd3881457b6da40016f/R/model.R#L804-L806

which that handles preparing the call to cmdstan, setting up the external processes, and then running cmdstan. It's also the object that report the run times (we don't wait for reading in the csv to report the time). The way I created `as_cmdstan_mcmc()` it avoids creating this object entirely and just uses the output from `read_cmdstan_csv()`. So to get time information with `as_cmdstan_mcmc()` we'd just need to add it to `read_cmdstan_csv()` (probably a good idea anyway).

**Comment by rok-cesnovar** [comment]
> we'd just need to add it to read_cmdstan_csv() (probably a good idea anyway). 

Agree. Will do that separately.

File: R/read_csv.R
Comment (rok-cesnovar) [reply R/read_csv.R]: I like them, but have another idea. What about `as_cmdstan_fit()` that does all three?

File: R/read_csv.R
Comment (rok-cesnovar) [reply R/read_csv.R:421]: What aobut CmdStanMCMCCSV, CmdStanMCMC_CSV? I am bad for names.

File: R/read_csv.R
Comment (rok-cesnovar) [reply R/read_csv.R:419]: I think having them here is fine.

File: R/read_csv.R
Comment (jgabry) [reply R/read_csv.R]: Good idea. I just changed it.

**Comment by jgabry** [comment]
I changed to `as_cmdstan_fit()`, good idea. I haven't changed the class names yet but I'm leaning towards `CmdStanMCMC_CSV`. That's definitely better than `CmdStanMCMC2`. 

Let's wait until  #414 is merged and then we can update this PR to include `time` before merging it.

**Comment by jgabry** [comment]
Ok I changed the class names to `CmdStanMCMC_CSV`, `CmdStanMLE_CSV`, `CmdStanVB_CSV` and I added the `time()` method for `CmdStanMCMC_CSV` after pulling in the changes from #414.

**Review by rok-cesnovar — APPROVED** [review]
Looks good to me. Thanks.

**Pull request merged**
